// @anei/common/build/events/base-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Listener = void 0;
// Everytime me extend Listener we need to provide an Event
class Listener {
    constructor(client) {
        this.ackWait = 5 * 1000; // 5sc we setup values here, don't need to be abstract
        this.client = client;
    }
    subscriptionOptions() {
        return this.client
            .subscriptionOptions()
            .setDeliverAllAvailable() // NATS resend all the events when the listener is up
            .setManualAckMode(true)
            .setAckWait(this.ackWait)
            .setDurableName(this.queueGroupName); // Create groups of transactions 
    }
    listen() {
        const subscription = this.client.subscribe(this.subject, this.queueGroupName, // use groups for more than one instance of a service to avoid to remove the durable names if a short disconnect
        this.subscriptionOptions());
        subscription.on('message', (msg) => {
            console.log(`Message received: ${this.subject} / ${this.queueGroupName}`);
            const parsedData = this.parseMessage(msg);
            this.onMessage(parsedData, msg);
        });
    }
    parseMessage(msg) {
        const data = msg.getData();
        return typeof data === 'string'
            ? JSON.parse(data)
            : JSON.parse(data.toString('utf8'));
    }
}
exports.Listener = Listener;
